### ISSUE TITLE
Add option to auto-create a path

### ISSUE THREAD

**Issue by sigxcpu76** [RedisJSON/RedisJSON#27]
Is it possible to add an option  to auto-create a JSON path in `JSON.SET` ?

It would be nice to set arbitrary values (e.g. counters for instrumentation) in a tree.

Right now there is no way to do this safely because creating the path using JSON can override existing values.

Example:

`JSON.SET test .key1.key2 10` - this fails because `.key1` does not exist
`JSON.SET test .key1 {}` - will create the key but if a `.key1.key3` exists it will be destroyed

**Comment by itamarhaber** [comment]
Hello Alexandru.

> Is it possible to add an option to auto-create a JSON path in JSON.SET ?

Anything is possible, it's just code :)

> Right now there is no way to do this safely ...

That's not entirely correct IIUC. What you could do is check for existence (with the `JSON.TYPE` command) and create if needed. This is less convenient and more expensive than your suggestion, but still works.

That said, the feature request is interesting and could be useful if only for simplifying the interaction with the document. I actually had something like it in a previous iteration of `JSON.SET`, i.e. implicit creation of missing nodes along the path, but I removed it. It didn't fit because it only operates on a very specific type of documents (nested dictionaries with terminal scalars, any referenced array elements along the path must exist and cannot be implicitly created), and handling all the edge cases in `JSON.SET` felt too complex. In addition, the existing `NX` and `XX` flags would have to be redifined/beefed up to express the all different combinations...

So my current thinking is that perhaps this functionality should be provided with a new command, e.g. `JSON.SETP` (P is for path and/or --parents like in mkdir). It will still support `NX`/`XX` with the same meaning for the last path node, but perhaps with `PNX`/`PXX` as well for the parents. It will require that all nodes along the path are either keys in dictionaries or array elements that exist. What do you think?

**Comment by nikolamakin** [comment]
when is time ok ?

**Comment by jacobbogers** [comment]
is this created?

**Comment by jpambrun** [comment]
The JSON.TYPE approach doesn't quite work for me because it leads to race conditions. Two threads can invoke JSON.TYPE at the same time, one quickly creates the path and adds a value and then the second, thinking the path is missing, overrides the path and content then add it's value. Thus, only the second value survives.

Am I missing something? Is there any other alternative?

**Comment by OmgImAlexis** [comment]
Yep this should have been fixed years ago as there's currently no way to use this without race conditions.

**Comment by DurandA** [comment]
> The JSON.TYPE approach doesn't quite work for me because it leads to race conditions. Two threads can invoke JSON.TYPE at the same time, one quickly creates the path and adds a value and then the second, thinking the path is missing, overrides the path and content then add it's value. Thus, only the second value survives.

Isn't this fixed by using a transaction?

**Comment by jpambrun** [comment]
I don't think it's possible to create transaction with conditions and branches in redis? How would that work?

**Comment by sosioo** [comment]
Redis automatically creating intermediate paths is by far the most needed feature in my opinion. Redis JSON not dealing with such basic thing makes so many uses cases complicated and costly. If it existed I could replace thousands of HASHES and SETs with one single JSON store. But that little single issue prevent me from doing so. Not worth it since it just needs as many commands to prevent or deal with that "can not follow further as XYZ does not exist".
I hope this will come soon in whatever form (upraded SET or new command)!. :)

**Comment by LiorKogan** [comment]
@sosioo thank you for your input.
We indeed plan to solve it in a future version for `ARRAPPEND`, `ARRINSERT`, `MERGE`, `NUMINCRBY`, `SET`, and `STRAPPEND`.

**Comment by jiahaoChen** [comment]
Hi, 
Is it workable? or has any other options can auto create a path?

**Comment by Nathan-Sokn** [comment]
You can use a method like this to auto-create nested paths of arbitrary depth. Definitely not pretty, but it works flawlessly until native support for this feature is added. 
```
   async jset(key: string, path: string, value: any, ttl?: number): Promise<void> {
        try {
            // Try direct JSON.SET first
            await this.jsonClient.json.set(key, path, value);
        } catch (error) {
            if (error.message.includes('new objects must be created at the root')) {
                // Get existing document or create empty one
                let document: any = {};
                try {
                    document = await this.jsonClient.json.get(key) || {};
                } catch {
                    // Key doesn't exist
                }
                
                // Parse JSONPath
                const pathParts = path.replace('$.', '').split('.');
                let current = document;
                
                // Create nested path
                for (let i = 0; i < pathParts.length - 1; i++) {
                    if (!current[pathParts[i]]) {
                        current[pathParts[i]] = {};
                    }
                    current = current[pathParts[i]];
                }
                
                // Set the value
                current[pathParts[pathParts.length - 1]] = value;
                
                // Save the entire document
                await this.jsonClient.json.set(key, '$', document);
            } else {
                throw error;
            }
        }
        
        if (ttl) {
            await this.jsonClient.expire(key, ttl);
        }
    }
```

**Comment by aradalvand** [comment]
No updates after 8 years?! This is a pretty basic feature, man.

**Comment by sigxcpu76** [comment]
https://valkey.io/commands/#json

**Comment by aradalvand** [comment]
@sigxcpu76 Why did you close the issue? RedisJSON doesn't have this feature yet, that link is for Valkey.

**Comment by sigxcpu76** [comment]
Because, as you said, no updates. I guess it is stale. World moved on :)
